### PR TITLE
feat(desktop): make queue reordering work with a mouse and a keyboard (#388)

### LINKED ISSUE #388: [Desktop] Add drag-and-drop queue management
Part of #376.

## Goal
Allow mouse drag-and-drop reordering of the desktop playback queue.

## Likely code areas
- `lib/features/player/`
- queue state/controller/providers
- queue list widgets
- player tests

## Requirements
- Dragging must reorder logical queue items without changing the currently playing track identity.
- Preserve repeat/shuffle semantics and queue indices after a move.
- Invalid drops must be harmless.
- Provide visible drag handles/feedback and a keyboard-accessible alternative.
- Reuse the existing queue model; do not add a second desktop queue implementation.

## Acceptance criteria
- [ ] Queue items can be reordered by drag-and-drop.
- [ ] Current track and playback position stay intact during reorder.
- [ ] Shuffle/repeat behavior remains correct.
- [ ] Rapid/multiple reorders do not corrupt queue state.
- [ ] Focused controller/widget tests cover edge cases.

### PR DESCRIPTION
**Pull request by TheZupZup**
Closes #388. Part of #376 (PR 4, desktop UX).

## What this is

The Up Next list was already reorderable, but the handle was built for a thumb: no cursor change, no lift while dragging, and no way to move a row at all without a pointer. This makes the same list behave like a desktop control.

- grab cursor and a hover tooltip on the handle, so a mouse user can see it is draggable before trying
- the dragged row lifts onto a shadow (`proxyDecorator`). Desktop pointers have no haptics and no long-press wind-up, so this is the only feedback that the drag actually took
- **Ctrl + ↑ / ↓** (Cmd on macOS, and the hint says so per platform) moves the focused row, with focus following the track and the sheet scrolling along, so the chord can be held to walk a row up or down the queue
- the same two moves as **Move up** / **Move down** semantics actions, and only the one that goes somewhere at either end of the list

No second desktop queue implementation: every route calls the existing `PlaybackController.reorderQueue` / `PlaybackQueue.reorderUpNext`, so drag, keyboard and screen reader all take the same path. The current track and playback are untouched, and shuffle/repeat keep the semantics they already had.

## Things worth a look in review

**The tooltip is hover-only** (`TooltipTriggerMode.manual`). The default long-press trigger puts a long-press recognizer in the arena next to the drag listener, and on a handle the press *is* the drag: the tooltip wins the arena and the row never lifts. This broke the two existing drag tests, which is how it was caught. Hover is the desktop trigger anyway, and the handle stays named for screen readers either way.

**The chord takes a modifier on purpose.** A bare arrow inside a scrolling sheet belongs to focus traversal and scrolling, so stealing it would trade one accessible behaviour for another.

**Focus is positional, and that is the hard part.** The up-next list is stateful because a pointer drag carries the row under the pointer while a keyboard move is a jump. The nodes are per position rather than per track (the same song can legitimately be queued twice), so anything that reorders the list changes which track a focused node belongs to. Three things fall out of that, all covered by tests:

- a held chord repeats faster than the queue reaches the sheet through its `StreamProvider`, so the second press arrives at a row still reporting its pre-move index. The list tracks where a walk started and where it actually landed, and drops both once a rebuild makes the indices truthful.
- focus can only land on a row the sliver built, so the moved row is scrolled into view before it takes focus. Without that, a long walk strands itself partway down and starts moving the wrong track.
- a pointer drag remaps the focused row through the move it just applied, covering both a drag of the focused row and a drag past it. It only fires when a handle already had focus, so a plain mouse drag never pulls focus into the list.

## Tests

Widget tests in `test/features/player/queue_sheet_test.dart`:

- the chord moves the focused row, and focus follows the track (a second press moves the same track)
- walking a row back up to the top
- a chord at either end is a no-op, not a clamp onto a neighbour or a push past the playing track
- a held chord (`sendKeyRepeatEvent`, no pump) keeps walking one track instead of undoing itself
- a walk of twelve rows down a 30-track queue moves one track twelve rows and leaves the other 28 alone
- a chord after a pointer drag moves the track that was dragged; a drag past a focused row carries that row along; a mouse drag alone never pulls focus into the list
- reordering a shuffled queue reorders the order actually playing, shuffle stays on
- rows offer the move actions, and only the ones that exist; a screen reader `Move down` actually reorders
- the handle reports a grab cursor on hover, and the hint names Cmd on macOS and Ctrl on Linux

Model tests in `test/core/models/playback_queue_test.dart`:

- repeated `reorderUpNext()` calls keep the queue whole (no lost, duplicated or half-moved entry)
- reordering while shuffled leaves `originalOrder` alone, so a later unshuffle still restores it

`docs/queue.md` covers the desktop and keyboard behaviour under Reordering, including that an overshot drag clamps to the end of the list while a keyboard move at an endpoint does nothing.

## Checks

Run locally against the pinned Flutter 3.44.7:

- `dart format --set-exit-if-changed .`
- `flutter analyze` (no issues)
- `flutter test` (4368 passing)

Android is untouched: this is the same widget on both platforms, and the whole suite passes.

## Review history

Codex raised five findings across two rounds. All five were real, all were reproduced with a failing test before being fixed, and all threads are resolved. Codex has since hit its account usage limit for code reviews, so later pushes will not get a review pass from it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01JGaoXaopecxf5D6Mv9Deb4

### CODE REVIEW

File: lib/features/player/widgets/queue_sheet.dart [hunk 1]
```diff
@@ -314,23 +293,149 @@ class _HistoryTile extends StatelessWidget {
...
+        // duplicate key (which would crash the reorderable list).
+        key: ValueKey<String>('queue-$index-${tracks[index].id}'),
+        track: tracks[index],
+        index: index,
+        count: tracks.length,
+        handleFocusNode: _focusNodeAt(index),
+        onPlay: () => ref.read(playbackControllerProvider).playFromQueue(index),
+        onRemove: () =>
+            ref.read(playbackControllerProvider).removeFromQueue(index),
+        onMoveBy: (int delta) => _move(index, index + delta, followFocus: true),
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/features/player/widgets/queue_sheet.dart]: Good catch, this is real. Reproduced it first with a test that sends a key down plus an auto-repeat with no frame in between, and it did exactly what you described: `[B, C, D]` back to `[B, C, D]` instead of B ending up last.

Worth noting the window is wider than "before the next frame rebuilds": the queue reaches the sheet through a `StreamProvider`, so the rebuild is a microtask plus a frame away, not just a frame.

Fixed in 7786c97 by tracking the walk instead of trusting the row: `_walkOrigin` is the stale index the un-rebuilt handles still report, `_walkLanded` is where the track actually is, and both are dropped in `didUpdateWidget` the moment a new track list makes the indices truthful again. The substitution is guarded on the originating row, so a press on some other row inside that same window is still taken at face value.

`a held chord keeps walking the same track` in `queue_sheet_test.dart` covers it, using `sendKeyRepeatEvent` with no pump so it stays in the window rather than pumping past it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/features/player/widgets/queue_sheet.dart [hunk 2]
```diff
@@ -314,23 +293,149 @@ class _HistoryTile extends StatelessWidget {
...
+  /// Puts keyboard focus back on the moved track's handle, after the frame that
+  /// rebuilds the list — before it, the destination row is still the old track.
+  /// A row scrolled out of view has no element to focus; skipping is fine,
+  /// the move itself already happened.
+  void _focusHandleAt(int index) {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      if (!mounted || index >= _handleFocusNodes.length) return;
+      final FocusNode node = _handleFocusNodes[index];
+      if (node.context == null) return;
+      node.requestFocus();
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/features/player/widgets/queue_sheet.dart:449]: Also real, and worse than a lost keystroke: a test walking a track twelve rows down a 30-track queue moved it eight and then started dragging its neighbours along, because focus stranded itself on the row a different track had slid into.

Fixed in 7786c97 with the scroll option you suggested. `_followWalkTo` now calls `Scrollable.ensureVisible` on the moved row before focusing it, with `keepVisibleAtEnd` / `keepVisibleAtStart` depending on direction so it only scrolls when the row would otherwise leave the viewport.

That also makes the unbuilt-destination case unreachable rather than just handled: the destination is always one row from a row that was just scrolled into view, so it stays inside the sliver's built range. The `node.context == null` return is now a genuine belt-and-braces path rather than something a long queue walks into.

I went with this over identity-keyed focus nodes because the same song can legitimately be queued twice, so queue-entry identity would need a synthetic key per entry, and per-position nodes are what make the destination node exist before the rebuild lands.

Covered by `a walk past the bottom of the sheet still moves one track`, which asserts both that the track ends up twelve rows down and that the other 28 are untouched.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/queue.md [hunk 3]
```diff
@@ -48,6 +48,20 @@ Drag an **upcoming** track by its handle to a new position. The current track is
 untouched and keeps playing — reordering only changes the order of what follows.
 History and the current track are not draggable.
 
+On desktop the handle behaves the way a desktop control should: the pointer
+turns into a grab cursor over it, hovering shows what it does, and the row lifts
+onto a shadow while it is being dragged so you can see what you picked up. A
+drop outside the list, or one that would land past either end, simply leaves the
+queue as it was.
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply docs/queue.md]: Right, the doc said the opposite of what `_dragToEdge(..., toEnd: true)` asserts. Fixed in 7786c97.

It now says an overshot drag drops the track at that end (framed as "never destructive", which is the bit that actually matters to a reader), and the no-op wording moved to where it belongs: a keyboard move at an endpoint, which genuinely does nothing.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/features/player/widgets/queue_sheet.dart [hunk 4]
```diff
@@ -314,23 +293,203 @@ class _HistoryTile extends StatelessWidget {
...
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Track> tracks = widget.tracks;
+    return SliverReorderableList(
+      itemCount: tracks.length,
+      onReorderItem: (int oldIndex, int newIndex) => _move(oldIndex, newIndex),
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/features/player/widgets/queue_sheet.dart]: Confirmed with a test before fixing, and the symptom was slightly worse than described: after dragging B to the end, the focus was left on the row C had slid into, which was row 0, so Ctrl+↑ hit the top of the list and did nothing at all. Silent no-op rather than a visible wrong move.

Fixed in 635afca. `_moveByPointer` finds the focused handle, applies the move, then remaps that row's index through the move it just made (`_positionAfterMove`). Doing the remap generally rather than special-casing the dragged row also covers the bystander case: dragging a row *past* a focused one shifts the focused row too, and the chord has to follow it.

It's guarded on a handle already having focus, so a plain mouse drag with nothing focused still doesn't pull focus into the list. There's a test pinning that specifically, since it would be an annoying regression to introduce while fixing this.

Three tests: `a chord after a pointer drag moves the track that was dragged`, `a drag past a focused row carries that row along`, and `a mouse drag alone never pulls focus into the queue`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/features/player/widgets/queue_sheet.dart [hunk 5]
```diff
@@ -374,6 +529,128 @@ class _UpNextTile extends StatelessWidget {
...
+            focusNode: focusNode,
+            child: Builder(
+              builder: (BuildContext context) {
+                final bool focused = Focus.of(context).hasFocus;
+                return MouseRegion(
+                  cursor: SystemMouseCursors.grab,
+                  child: ReorderableDragStartListener(
+                    index: index,
+                    child: Tooltip(
+                      message: 'Reorder (drag, or Ctrl + ↑ / ↓)',
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/features/player/widgets/queue_sheet.dart]: Fair, the hint contradicted both the docs and the registered activators. Fixed in 635afca: the tooltip now reads the modifier off `Theme.of(context).platform`, so macOS gets Cmd and everything else gets Ctrl.

I kept both activators registered rather than making the bindings platform-specific. An unused one costs nothing, and on macOS the OS swallows Ctrl+Arrow before it reaches us anyway, so there's nothing to gate. Only the wording needed to change.

Covered both directions (`the hover hint names the modifier that works on this platform` and `the hover hint says Ctrl on Linux`) so the macOS branch can't quietly rot, given Linthra doesn't ship a macOS build today.

---
_Generated by [Claude Code](https://claude.ai/code)_
